implement more comprehensive redis option passthrough

marknach opened the pull request:
## Hey, I just made a Pull Request!

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->
#29019 
I added the ability to pass through additional redis configuration for the core Cache service.  These should mirror the objects defined in noderedis for cluster config [here](https://github.com/redis/node-redis/blob/master/packages/client/lib/cluster/index.ts#L32) and non cluster [here](https://github.com/redis/node-redis/blob/master/packages/client/lib/client/index.ts#L21)

Sample config for clustered redis:
```yaml
  cache: 
    store: redis
    redis: 
      cluster:
        rootNodes:
          - url: rediss://${REDIS_HOST}:${REDIS_PORT}
        defaults:
          socket:
            tls: true
          password: ${REDIS_PASSWORD}
```

Sample config for non-clustered redis with using the option passthrough over the connection string
```yaml
  cache: 
    store: redis
    redis:
      client: 
        url: 'redis://localhost:6379'
        password: ${REDIS_PASSWORD}
``` 

I'm intentionally using `config.get()` vs destucturing the configuration passed through to allow flexibility in any supported configuration by keyv / noderedis.  Open to changing this if yall feel strongly about it.
#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

marknach commented on packages/backend-defaults/src/entrypoints/cache/CacheManager.ts:
```diff
@@ -89,22 +95,81 @@ export class CacheManager {
+  /**
+   * Parse store-specific options from configuration.
+   *
+   * @param store - The cache store type ('redis', 'memcache', or 'memory')
```
I had originally written some generic wrappers for other cache types - I could make this doc more apparent we're only accepting additional options for redis?

marknach replied:
Do you have any thoughts here? if we wanted to evaluate inside the `createRedisStoreFactory` we would still need a reference to the root config, or subconfig. 

I don't see a clean way to isolate this without spreading out where we touch the config

freben commented on packages/backend-defaults/src/entrypoints/cache/CacheManager.ts:
```diff
@@ -89,22 +95,81 @@ export class CacheManager {
+    redisOptions.client = redisConfig.getOptional('client');
+
+    if (redisConfig.has('cluster')) {
+      const clusterConfig = redisConfig.getConfig('cluster');
```
Whenever you add config, the config.d.ts file in the root of the package needs to be updated accordingly. In particular, if some fields can be secrets in here, they should be marked with `@visibility secret` so that they don't leak.

freben reviewed (commented):
Thanks for making this!

In general - hmmm how many options are we talking about here? I don't mind the `cluster` subkey to clearly mark that here are the cluster settings. But for the rest, are there really so many options that it's worth just giving up and essentially going `object` type on them? :)

If you think it's warranted, then sure, maybe. But the reason I'm asking is because if we make it completely opaque and unvalidated, we're very strongly tying ourselves not just to this exact redis client library, but to this VERSION of the client library too and when we update in the future (or switch out because it became unmaintained or so) it gets harder because we have no idea what people put in here and whether it'll work with that updated version.

marknach commented on packages/backend-defaults/src/entrypoints/cache/CacheManager.ts:
```diff
@@ -140,13 +206,23 @@ export class CacheManager {
     const KeyvRedis = require('@keyv/redis').default;
-    const stores: Record<string, typeof KeyvRedis> = {};
+    const { createCluster } = require('@keyv/redis');
+    const stores: Record<string, any> = {};
```
not sure where this came from 😬 changing it back

marknach commented:
@freben thanks for the review - i have:
* Cleaned up and limit the types for the configuration
* Updated config.d.ts appropriately 
* fixed bug in unit test you pointed out + corrected an "any" type i added

Two things unresolved / seeking feedback:
* https://github.com/backstage/backstage/pull/29162#discussion_r1997776949
* https://github.com/backstage/backstage/pull/29162/files#diff-3b45dbb2aa9a6a351eed560b0388e73b659ab55bc693efabecbb307b7bbe9b0aR600-R607 - I mimicked what i saw for knex/db configuration. I could enumerate more here but there are a few large nested options within here that are fairly commonly used

EricQi7 commented:
Hi marknach, I'm currently trying to configure AWS serverless redis on backstage using the following setup:

backend:
  cache: 
    store: redis
    connection: '[AWS_serverless_redis_endpoint]'
    redis: 
      cluster:
        rootNodes:
          - url:  '[AWS_serverless_redis_endpoint]'
        defaults:
          socket:
            tls: true
I also attempted to configure redis with cluster mode enabled as follows:

backend:
  cache: 
    store: redis
    connection: '[AWS_cluster_mode_redis_endpoint]'
    redis: 
      cluster:
        rootNodes:
          - url:  '[AWS_redis_root_node1_endpoint]'
          - url:  '[AWS_redis_root_node2_endpoint]'
          - url:  '[AWS_redis_root_node3_endpoint]'
        defaults:
          socket:
            tls: true
However, I keep encountering the following error:

techdocs error Request failed with status 500 Cannot read properties of undefined (reading 'replicas') type="errorHandler" stack="TypeError: Cannot read properties of undefined (reading 'replicas')\n    at RedisClusterSlots.getSlotRandomNode (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:118:19)\n    at RedisClusterSlots.getClient (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/cluster-slots.js:110:37)\n    at RedisCluster._RedisCluster_execute (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/index.js:217:79)\n    at RedisCluster.sendCommand (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/index.js:129:98)\n    at RedisCluster.commandsExecutor (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/cluster/index.js:126:75)\n    at BaseClass.<computed> [as get] (/Users/Documents/Repository/node_modules/@redis/client/dist/lib/commander.js:8:29)\n    at KeyvRedis.get (/Users/Documents/Repository/node_modules/@backstage/backend-defaults/node_modules/@keyv/redis/dist/index.cjs:262:32)\n    at async Keyv.get (/Users/Documents/Repository/node_modules/keyv/dist/index.cjs:478:21)\n    at async DefaultCacheClient.get (/Users/Documents/Repository/node_modules/@backstage/backend-defaults/src/entrypoints/cache/CacheClient.ts:52:19)\n    at async CachedEntityLoader.getFromCache (/Users/Documents/Repository/node_modules/@backstage/plugin-techdocs-backend/src/service/CachedEntityLoader.ts:63:13)\n    at async CachedEntityLoader.load (/Users/Documents/Repository/node_modules/@backstage/plugin-techdocs-backend/src/service/CachedEntityLoader.ts:45:18)\n    at async <anonymous> (/Users/Documents/Repository/node_modules/@backstage/plugin-techdocs-backend/src/service/router.ts:161:20)"

**The AWS standalone redis is successfully connected. But I was wondering if you've tested this configuration with AWS serverless and cluster mode redis, and if so, could you kindly advise on what might be causing this issue? Any guidance or support would be greatly appreciated.**

**Thank you very much for your time and help!**
